### PR TITLE
add smartspaces plugin

### PR DESCRIPTION
**Pull request by mbkma**
Ported from https://gitlab.gnome.org/GNOME/gedit-plugins
Do not merge before https://github.com/mate-desktop/pluma/pull/615

### CODE REVIEW

**Review by rbuj — COMMENTED** [review]
Translations of the plugin file will not be captured:
```diff
diff --git a/po/POTFILES.in b/po/POTFILES.in
index 5ca20e3..6bc4bf0 100644
--- a/po/POTFILES.in
+++ b/po/POTFILES.in
@@ -6,6 +6,7 @@ plugins/bookmarks/pluma-bookmarks.metainfo.xml.in.in
 plugins/codecomment/codecomment.plugin.desktop.in.in
 plugins/codecomment/codecomment.py
 plugins/codecomment/pluma-codecomment.metainfo.xml.in.in
+plugins/smartspaces/smartspaces.plugin.desktop.in.in
 plugins/synctex/pluma-synctex.metainfo.xml.in.in
 plugins/synctex/synctex.plugin.desktop.in.in
 plugins/synctex/synctex/synctex.py
```

File: plugins/smartspaces/Makefile.am [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+plugin_DATA = $(plugin_in_files:.desktop.in=)
+
+$(plugin_DATA): $(plugin_in_files)
+	$(AM_V_GEN) $(MSGFMT) --keyword=Name --keyword=Description --desktop --template $< -d $(top_srcdir)/po -o $@
+
+EXTRA_DIST = $(plugin_in_files)
+
+CLEANFILES = $(plugin_DATA)
+
+DISTCLEANFILES = $(plugin_DATA)
```
Comment (rbuj) [inline plugins/smartspaces/Makefile.am]: the `distclean` target calls `clean` target, so you don't need to add the CLEANFILES file list in DISTCLEANFILES.

Comment (rbuj) [reply plugins/smartspaces/Makefile.am]: You can set `DISTCLEANFILES = $(plugin_in_files)` instead.

Comment (mbkma) [reply plugins/smartspaces/Makefile.am]: done

File: plugins/smartspaces/Makefile.am [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+plugin_PYTHON = \
+	smartspaces.py
+
+plugin_in_files = smartspaces.plugin.desktop.in
+plugin_DATA = $(plugin_in_files:.desktop.in=)
+
+$(plugin_DATA): $(plugin_in_files)
+	$(AM_V_GEN) $(MSGFMT) --keyword=Name --keyword=Description --desktop --template $< -d $(top_srcdir)/po -o $@
+
+EXTRA_DIST = $(plugin_in_files)
```
Comment (rbuj) [inline plugins/smartspaces/Makefile.am]: You should update EXTRADIST:
```diff
diff --git a/plugins/smartspaces/Makefile.am b/plugins/smartspaces/Makefile.am
index 3606504..453e5ed 100644
--- a/plugins/smartspaces/Makefile.am
+++ b/plugins/smartspaces/Makefile.am
@@ -3,13 +3,14 @@ plugindir = $(PLUMA_PLUGINS_LIBS_DIR)
 plugin_PYTHON = \
        smartspaces.py
 
-plugin_in_files = smartspaces.plugin.desktop.in
+plugin_in_in_files = smartspaces.plugin.desktop.in.in
+plugin_in_files = $(plugin_in_in_files:.in=)
 plugin_DATA = $(plugin_in_files:.desktop.in=)
 
 $(plugin_DATA): $(plugin_in_files)
        $(AM_V_GEN) $(MSGFMT) --keyword=Name --keyword=Description --desktop --template $< -d $(top_srcdir)/po -o $@
 
-EXTRA_DIST = $(plugin_in_files)
+EXTRA_DIST = $(plugin_in_in_files)
 
 CLEANFILES = $(plugin_DATA)
 
```

**Comment by rbuj** [comment]
You need to update the required version of pluma after releasing a new version / tarball:
https://github.com/mate-desktop/pluma-plugins/blob/0b4add59a4685d46aa6dbc199bef1a9f42454d0f/configure.ac#L57
Next, you need to download and build pluma from the new tarball since it has not been published yet.
Otherwise, the build will fail as it won't find the new interfaces.

**Comment by raveit65** [comment]
Just rebased with master to use travis-CI.

**Comment by raveit65** [comment]
Rebased with master to build with pluma-1.25.2.

**Comment by rbuj** [comment]
You need to bump the version of pluma:
```diff
diff --git a/configure.ac b/configure.ac
index 5eb32d0..e743b3a 100644
--- a/configure.ac
+++ b/configure.ac
@@ -54,7 +54,7 @@ PKG_CHECK_MODULES(PLUMA, [
        gtksourceview-3.0 >= 3.9.2
        libpeas-1.0 >= 1.7.0
        libpeas-gtk-1.0 >= 1.7.0
-       pluma >= 1.24.0
+       pluma >= 1.25.2
 ])
 
 dnl ================================================================
```

**Comment by mbkma** [comment]
@rbuj @raveit65 sorry, I pushed the pluma version requirement to master, but I meant to push it here...

**Comment by rbuj** [comment]
~~It seems this plugin interferes with the editor setting "Insert spaces instead of tabs", when the tab key is pressed it writes the tab character even if this plugin is enabled.~~

**Review by rbuj — APPROVED** [review]
LGTM

**Comment by rbuj** [comment]
Test:

1. Select "Insert spaces instead of tabs" on Editor preferences
2. Set "Tab width" to 4 spaces
2. Enable "Smart Spaces" plugin
3. Write a word preceded by several whitespaces
4. Move the cursor before the word
5. Press Backspace key.
 
It should remove 4 spaces each time you press the Backspace key, you can also press Shift+Tab

**Pull request merged**